crates/typed-store/src/sally/mod.rs

```diff
@@ -0,0 +1,588 @@
+//! use typed_store_derive::SallyDB;
+//! use typed_store::sally::SallyColumn;
+//! use typed_store::traits::TypedStoreDebug;
+//! use crate::typed_store::Map;
+//!
+//! // `ExampleTable` is a sally db instance where each column is first initialized with TestDB
+//! // (btree map) backend and later switched to a RocksDB column family
+//!
+//! #[derive(SallyDB)]
+//! pub struct ExampleTable {
+//!   col1: SallyColumn<String, String>,
+//!   col2: SallyColumn<i32, String>,
+//! }
+//!
+//! fn update_table(table: &ExampleTable) {
+//!     table.col1.insert(&"key1".to_string(), &"value1".to_string());
+//!     table.col2.insert(&1, &"test".to_string());
+//!     // create a write batch and do atomic commit
+//!     let keys_vals = (1..100).map(|i| (i, i.to_string()));
+//!     let mut wb = table.col1.batch();
+//!     wb = wb.insert_batch(&table.col2, keys_vals).expect("Failed to batch insert");
+//!     wb = wb.delete_range(&table.col2, &50, &100).expect("Failed to batch delete");
+//!     wb.write();
+//! }
+//!
+//! #[tokio::main]
+//! async fn main() -> Result<(), TypedStoreError> {
+//!     // use a btree map backend first
+//!     let mut table = ExampleTable::init(SallyDBOptions::TestDB);
+//!     update_table(&table);
+//!     let primary_path = tempfile::tempdir().expect("Failed to open path").into_path();
+//!     // switch to rocksdb backend
+//!     table = ExampleTable::init(SallyDBOptions::RocksDB((primary_path, RocksDBAccessType::Primary, None, None)));
+//!     update_table(&table);
+//!     Ok(())
+//! }
+//! ```
+use crate::{
+    rocks::{
+        default_db_options, keys::Keys, values::Values, DBBatch, DBMap, DBOptions,
+        RocksDBAccessType, TypedStoreError,
+    },
+    test_db::{TestDB, TestDBIter, TestDBKeys, TestDBValues, TestDBWriteBatch},
+    traits::Map,
+};
+
+use crate::rocks::iter::Iter as RocksDBIter;
+use crate::rocks::DBMapTableConfigMap;
+use collectable::TryExtend;
+use rocksdb::Options;
+use serde::{de::DeserializeOwned, Serialize};
+use std::borrow::Borrow;
+use std::{collections::BTreeMap, path::PathBuf};
+
+pub enum SallyRunMode {
+    // Whether Sally should use its own memtable and wal for read/write or just fallback to
+    // reading/writing directly from the backend db. When columns in the db are backed by different
+    // backend stores, we should never use `FallbackToDB` as that would lose atomicity,
+    // transactions and db recovery
+    FallbackToDB,
+}
+
+pub struct SallyConfig {
+    pub mode: SallyRunMode,
+}
+
+impl Default for SallyConfig {
+    fn default() -> Self {
+        Self {
+            mode: SallyRunMode::FallbackToDB,
+        }
+    }
+}
+
+/// A Sally column could be anything that implements key value interface. We will eventually have
+/// Sally serve read/writes using its own memtable and wal when columns in the db are backend by more then
+/// one backend store (e.g different rocksdb instances and/or distributed key value stores)
+pub enum SallyColumn<K, V> {
```

Comment:
I thought of going in this direction, and now I can't remember exactly why I didn't pursue it. Let me give it  a try using traits

Reply:
After thinking some more, I realized the main reason I did not want to have different structs implementing a trait with something like:
```
trait SallyColumn<K, V> {
   async fn get();
   async fn multi_get();
   ...
}
```
was primarily because the usage then looks like:
```
struct MyTable {
   col1: Box<dyn SallyColumn<K, V>>;
   col2: Box<dyn SallyColumn<K, V>>;
}
impl MyTable {
    fn open(config: &SallyConfig) -> Self {
        // init rocksdb or init some other db based on the configs 
        MyTable {
           col1,
           col2
        }
    }
}
```
is we end up with individual db implementations of `SallyColumn<K, V>` (rocksdb, btreemap, etc) which are indeed complete but does not give us ability to build a layer of memtable or wal on top (on top of individual implementations) which can work with underlying implementations (when the time comes and we want to start using our own wal + memtable). Having an enum which can have its own implementation of `get`, `multiget`, etc gives us that room to implement those other things in future.

Reply:
Got it, thanks!